Refactoring issue 1 10

belkassaby opened the pull request.

belkassaby committed:
Create DownloadableResource

*Fix issue #2

belkassaby committed:
Fix #1 #3 #4 #5 #6 #7 #8 #9 #10

*Refactor the project name to BioDownload
*Create BioDownloaderBuilder #1 with all necessary methods to download
resources

belkassaby committed:
Fix POM file and disable tests

*put junit version
*define build goal
*disable the tests

belkassaby committed:
#2 Replace string name by Path and string url by URL

*Better handle URL and Paths
*add more logging
*add more documentation

pnrobinson reviewed (approved):
Looks great!
It might be useful to test that createURL does not return null for any of our predefined Resources.
Probably the package name should be
org.monarchinitiative.biodownload to make things easier with maven central as discussed.

belkassaby committed:
Change package name from org.jax to org.monarchinitiative

belkassaby committed:
Change package name from org.jax to org.monarchinitiative

belkassaby commented:
Thanks for the review! I just committed the change of package name. Also for the url being null, I'm checking it in the `validate` method of `BioDownloaderBuilder` line 206.

belkassaby committed:
Remove old files in org.jax package and add more info when url is null

belkassaby committed:
Merge branch 'Refactoring_issue_1_10' of git@github.com:TheJacksonLaboratory/biodownload.git into Refactoring_issue_1_10

ielis commented on src/main/java/org/monarchinitiative/biodownload/BioDownloaderBuilder.java:
```diff
@@ -0,0 +1,224 @@
+    private boolean overwrite = false;
+
+
+    public BioDownloaderBuilder(String destinationPath) {
```
I think it is better to make this method to accept `Path destinationPath`. Otherwise, the user may provide just any string and the `obodownload` needs to handle one more potential bug, while construction of a valid `Path destinationPath` in the client environment is not (I think) responsibility of `obodownload`.

belkassaby replied:
resolved in e89e320c

ielis commented on src/main/java/org/monarchinitiative/biodownload/BioDownloaderBuilder.java:
```diff
@@ -0,0 +1,224 @@
+     * @param urlStr URL string of source file
+     * @return a builder instance
+     */
+    public BioDownloaderBuilder custom(String name, String urlStr) {
```
I think it may be better to ask for `URL urlStr` here, otherwise we must handle the possibility of a malformed URL. I think this should not be responsibility of `obodownload` and forcing the user to handle the `MalformedURLException` may be beneficial.

belkassaby replied:
resolved in e89e320c

ielis commented on src/main/java/org/monarchinitiative/biodownload/BioDownloaderBuilder.java:
```diff
@@ -0,0 +1,224 @@
+     * @return a builder instance
+     */
+    public BioDownloaderBuilder custom(String name, String urlStr) {
+        URL url = DownloadableResource.createURL(urlStr);
```
This is something we should discuss. On the line below, a `DownloadableResource` with `null` is potentially added into `resources`, which unnecessarily complicates the logic. 
My preference is not having null-containing value objects in the app, unless the null represents an optional attribute. However, I think `DownloadableResource` with `null` is useless, it only complicates the logic. I would prefer `DownloadableResource.url` to be non-nullable.

belkassaby replied:
resolved in e89e320

ielis commented on src/main/java/org/monarchinitiative/biodownload/BioDownloaderBuilder.java:
```diff
@@ -0,0 +1,224 @@
+        if (resources.size() == 0) {
+            sb.append("A name and a URL need to be included. Please pick one of the available options or add your custom name and URL.");
+        }
+        for (DownloadableResource resource : resources) {
```
This check is appropriate right now, but no action is taken to remove the (I think) useless null-containing `DownloadableResource`.

ielis replied:
Actually I see now that the presence of `null` will throw `IllegalStateException`.

ielis commented on src/main/java/org/monarchinitiative/biodownload/BioDownloaderBuilder.java:
```diff
@@ -0,0 +1,224 @@
+     * @throws IllegalStateException If the validation fails we throw an {@link IllegalStateException}
+     */
+    private void validate() throws IllegalStateException {
+        StringBuilder sb = new StringBuilder();
```
I like this pattern for reporting all issues instead of just the first one.

ielis commented on src/main/java/org/monarchinitiative/biodownload/BioDownloaderImpl.java:
```diff
@@ -0,0 +1,75 @@
+            File file = downloadFileIfNeeded(downloadDirectory.resolve(resource.getName()), resource.getUrl());
+            downloadedFiles.add(file);
+            numberOfFiles++;
+            System.out.printf("[INFO] Downloaded \"%s\" file to \"%s\" (%d files were previously downloaded)\n",
```
`System.out.printf` calls are burden of the past, when our team did not use `slf4j` for logging. Please feel free to replace any direct `sout`/`serr` calls with an appropriate logging statement.

belkassaby replied:
resolved in 10b23b9

ielis commented on src/main/java/org/monarchinitiative/biodownload/BioDownloaderImpl.java:
```diff
@@ -0,0 +1,75 @@
+
+        for (DownloadableResource resource : resources) {
+            File file = downloadFileIfNeeded(downloadDirectory.resolve(resource.getName()), resource.getUrl());
+            downloadedFiles.add(file);
```
This can be `null`. My preference is to not return `null`s if possible. If there is no better option, wrapping in `Optional` improves clarity without having to mention *nullity* in javadocs which, unfortunately, not all people read.

belkassaby replied:
resolved in 99629b7

ielis commented on src/main/java/org/monarchinitiative/biodownload/BioDownloaderImpl.java:
```diff
@@ -0,0 +1,75 @@
+            logger.trace(String.format("Cowardly refusing to download %s since we found it at %s",
+                    f.getName(),
+                    f.getAbsolutePath()));
+            return null;
```
We can wrap this in `Optional`

belkassaby replied:
resolved in 99629b7

ielis commented on src/main/java/org/monarchinitiative/biodownload/DownloadableResource.java:
```diff
@@ -0,0 +1,81 @@
+     */
+    public DownloadableResource(String name, URL url) {
+        this.name = name;
+        this.url = url;
```
My preference is to ensure that fields of POJOs/value objects are not null. If you think it can be helpful, a way how to ensure non-nullity is to add something like 
```
this.url = Objects.requireNonNull(url, "Url must not be null")
``` 
here.
The same applies for `this.name`

belkassaby replied:
resolved in e89e320

ielis commented on src/main/java/org/monarchinitiative/biodownload/DownloadableResource.java:
```diff
@@ -0,0 +1,81 @@
+        try {
+            return createURLExc(urlStr);
+        } catch(MalformedURLException mue) {
+            logger.error(String.format("Malformed URL for %s", urlStr));
```
`String.format` is redundant in logging statements. It is computationally cheaper to use
```
logger.error("Malformed URL for {}", urlStr);
```
because the statement is not generated if the logging level is below the logging threshold. E.g. the user configured the loggers to only show `>=WARN`, and the message is `DEBUG`.
In the case above, the `String.format(...)` is called no matter what.

belkassaby replied:
resolved in commit 10b23b9

ielis commented on src/main/java/org/monarchinitiative/biodownload/DownloadableResource.java:
```diff
@@ -0,0 +1,81 @@
+     */
+    public static URL createURLExc(String urlStr) throws MalformedURLException{
+        URL url = new URL(urlStr);
+        logger.debug("Created url from " + urlStr + ": "+ url);
```
We can use
```
logger.debug("Created url from {}: {}", urlStr, url);
```
here too.

belkassaby replied:
resolved in commit 10b23b9

ielis commented on src/main/java/org/monarchinitiative/biodownload/DownloadableResource.java:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Class used to store a downloadable resource and with all the resources objects that can be downloaded.
+ */
+public class DownloadableResource {
```
We can add `equals()`, `hashCode`, and `toString` methods here.

belkassaby replied:
resolved in 5c6e06a

ielis commented on src/test/java/org/monarchinitiative/biodownload/FileDownloaderImplTest.java:
```diff
@@ -0,0 +1,29 @@
+    private IBioDownloader bioDownloader;
+
+    @BeforeEach
+    void init() {
```
This should be `public` too, otherwise we may run into issues with inaccessibility of the method after we switch to JPMS (Jigsaw).

belkassaby replied:
resolved in 5c6e06a

ielis reviewed (changes requested):
Hi @belkassaby ,
thank you for the PR - I went through the commits and I added my notes. I'd like us to reach a consensus in usage of `null`s. I am also a fan of delegating some tasks to the client code, e.g. handling `MalformedURLException`, creating the `Path` or `File`. Please let's consider these points - these are the reason why I'm asking for changes.

The rest of my notes are probably a matter of personal preference.

Thank you and talk to you soon!
Daniel

belkassaby committed:
Add test files

belkassaby committed:
Delegate URL and Path creation to client

*have URL and Path as parameters instead of having to create them
ourselves and have to handle the corresponding possible exceptions
*do not accept null url & name
*create some new tests

belkassaby committed:
Update/fix tests

belkassaby committed:
Better manage when file returned is null

*do not store a null file object into the returned List

belkassaby committed:
Wrap null file into Optional

belkassaby committed:
Add missing hashcode, equals and toString to DownloadableResource

belkassaby committed:
Replace sysinfo with logger

belkassaby commented:
> Hi @belkassaby , thank you for the PR - I went through the commits and I added my notes. I'd like us to reach a consensus in usage of `null`s. I am also a fan of delegating some tasks to the client code, e.g. handling `MalformedURLException`, creating the `Path` or `File`. Please let's consider these points - these are the reason why I'm asking for changes.
> 
> The rest of my notes are probably a matter of personal preference.
> 
> Thank you and talk to you soon! Daniel

Thanks so much for the review @ielis .
 I commented under each with a link to the corresponding commit addressing the issue.

belkassaby committed:
return the correct string value of the url in toString

pnrobinson reviewed (changes requested):
Looks great. But I am not convinced this sort of thing is a good idea
public BioDownloaderBuilder go() {
        return goJson().goObo();
    }
because any given user will only want one of these items. This is likely to be confusing, and I would suggest we stick with just goObo, goJson, goOwl etc.

belkassaby commented:
> Looks great. But I am not convinced this sort of thing is a good idea public BioDownloaderBuilder go() { return goJson().goObo(); } because any given user will only want one of these items. This is likely to be confusing, and I would suggest we stick with just goObo, goJson, goOwl etc.

@pnrobinson Should I remove the all() method too?

pnrobinson commented:
Yes, I do not think there is any use case for it.

belkassaby committed:
Remove unnecessary methods from builder

pnrobinson reviewed (approved):
+1

belkassaby commented:
@pnrobinson @ielis  I think I've addressed all your comments. Let me know if you think I can merge into master. Then I'll create another branch to address the issue #13

Commit:
Make the static members of `DownloadableResource` final

Commit:
Make the methods for URL creation private

ielis reviewed (approved):
Looks great, thank you!

belkassaby commented:
Thanks to the both of you!
